File: python/34. Search for a Range.py
class Solution(object):
    def searchRange(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        left = self.helper(nums, target)
        if left < len(nums) and nums[left] == target:
            right = self.helper(nums, target + 1) - 1
            return [left, right]
        return [-1, -1]
    def helper(self, nums, target):
        l, r = 0, len(nums)
        while l < r:
            mid = (l + r) / 2
            if nums[mid] >= target:
                r = mid
            else:
                l = mid + 1
        return r

File: cpp/1-50/34. Search for a Range.cpp
class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target) 
    {
        vector<int> res(2, -1);
        if(nums.empty())
            return res;
        res[0] = serachLeft(nums, target);
        res[1] = serachRight(nums, target);
        return res;
    }
    int serachLeft(vector<int>& nums, int target)
    {
        int low = 0, high = nums.size() - 1;
        while(low <= high)
        {
            int mid = (low + high) / 2;
            if(nums[mid] < target)
                low = mid + 1;
            else
                high = mid - 1;
        }
        return nums[low] == target ? low: -1;
    }
    int serachRight(vector<int>& nums, int target)
    {
        int low = 0, high = nums.size() - 1;
        while(low <= high){
            int mid = (low + high) / 2;
            if(nums[mid] > target)
                high = mid - 1;
            else
                low = mid + 1;
        }
        return nums[high] == target? high: -1;
    }
};